Update djwt to version 2.0

asantos00 opened the pull request:
This PR updates `djwt` to version 2.0. Making it work flawlessly with the current deno version (1.6.0)

## Breaking changes
- onFailure and onSuccess callbacks now get an `Error` or a `Payload` (from `djwt`) respectively
- Middleware now receives an `AlgorithmInput` instead of `Algorithm` (exported from `djwt`)

## New exports from `mod.ts`
- `Payload` - the value injected in `onSuccess` callback

## Major changes:
- Refactor middleware method to (try) to make error cases a little more clear, also tried to avoid mutation (and variables like `isAuthorized` and so on)
- Proxy errors from `djwt` to `onFailure` callback to more comprehensive errors
- Adds integration tests that spawn an instance of `oak` with the middleware and run tests against it 
- Adds unit tests for `onSuccess` and `onFailure`
- Github action needs to be changed to run `make test` instead of `deno test`, so that it includes the new permissions needed for the integration tests.
- Fixes a "bug" (please confirm if it's actually a bug) where expired tokens weren't making authentication fail.

I hope I can help you with this and we can move this forward, let me know if there's something more we can do!

asantos00 committed:
Use export type

asantos00 committed:
test: refactor to improve clarity of error cases

asantos00 committed:
test: add expired token test

asantos00 committed:
test: add integration tests with oak

asantos00 committed:
test: comment expired token tests

asantos00 committed:
chore: upgrade djwt to v2.0

BREAKING CHANGE: the middleware API changed

- Due to the changes made in djwt, the middleware doesn't support critHandlers anymore.
- onSuccess and onFailure method signatures have changed callbacks were also changed.
- Old exports from djwt are not exported anymore, only Payload and getNumeric date

asantos00 committed:
test: uncomment expired date token test

halvardssm commented on deps.ts:
```diff
@@ -1,26 +1,27 @@
   assert,
+  assertEquals,
   assertThrowsAsync,
 } from "https://deno.land/std@0.74.0/testing/asserts.ts";
```
While we are at it, let's bump the dependencies as well 💪🏻 

```suggestion
} from "https://deno.land/std@0.82.0/testing/asserts.ts";
```

halvardssm commented on deps.ts:
```diff
@@ -1,26 +1,27 @@
-export {
-  Context,
+
+export type {
```
Is there a reason for why you split these imports up in types and non types?

asantos00 replied:
Yes! After the `--isolatedModules` flag was enabled by default on Deno (after version 1.5) this throws errors. It's now mandatory to use import and export types.

halvardssm replied:
Ah, yes, true... It has been a minor version since I dealt with Deno 😅

halvardssm commented on test.ts:
```diff
@@ -44,27 +37,47 @@ const mockNext = () => {
+    },
+  },
+  {
+    name: "Failure with expiretd token",
```
```suggestion
    name: "Failure with expired token",
```

halvardssm commented on src/jwtMiddleware.ts:
```diff
@@ -88,53 +89,57 @@ export const jwtMiddleware = <
-            onSuccess && onSuccess(ctx, validatedJwt);
-          }
-        }
+    const onUnauthorized = async (
```
Just have to say, this is a lot cleaner than my initial implementation 💪🏻

halvardssm commented on src/jwtMiddleware.ts:
```diff
@@ -88,53 +89,57 @@ export const jwtMiddleware = <
-    await next();
+    // No Authorization header
+    if (!ctx.request.headers.has("Authorization")) {
+      await onUnauthorized(new Error("Authorization header is not present"));
```
Let's use the error messages object here. Can you add "Authorization header is not present" to the object so that we can expose customization for the user?

halvardssm commented on src/jwtMiddleware.ts:
```diff
@@ -88,53 +89,57 @@ export const jwtMiddleware = <
+    // Authorization header has no Bearer or no token
+    const authHeader = ctx.request.headers.get("Authorization")!;
+    if (!authHeader.startsWith("Bearer ") || authHeader.length <= 7) {
+      await onUnauthorized(new Error("Invalid Authorization header"));
```
same as above

halvardssm commented on src/jwtMiddleware.ts:
```diff
@@ -88,53 +89,57 @@ export const jwtMiddleware = <
+    }
+
+    const jwt = authHeader.slice(7);
+    try {
+      onSuccess(ctx, await verify(jwt, key, algorithm));
+      await next();
+    } catch (e) {
+      await onUnauthorized(e);
+    }
```
Let's expand `onUnauthorized` with a second param (can be an optional boolean). Reason for this is that we can isolate the djwt errors from the middleware errors, and can return the djwt error instead of a predefined error as there is no easy way to distinguish the two. 

```suggestion
    try {
      onSuccess(ctx, await verify(jwt, key, algorithm));
      await next();
    } catch (e) {
      await onUnauthorized(e, true);
    }
```

asantos00 replied:
I tend not to be a big fan of passing flags as second arguments. Here I think it's not too bad since it's an internal thing so I'm fine with it 👍 

What type of different handling were you thinking of doing? Were you thinking of sending this flag into the `onFailure` callback, for instance?

halvardssm replied:
see the comment bellow

halvardssm reviewed (changes requested):
Hi @asantos00 ! Amazing job with the update! I added some notes, but after these are completed, we should be ready to merge 💪🏻 You can edit the [ci file](https://github.com/halvardssm/oak-middleware-jwt/blob/master/.github/workflows/ci.yml#L29) directly to use the make command, but be sure to update the [Deno version](https://github.com/halvardssm/oak-middleware-jwt/blob/master/.github/workflows/ci.yml#L4) as well since it is a bit outdated 😉  

EDIT: 
Also, don't forget to update the readme with the changes 💪🏻

asantos00 committed:
Update deps.ts std-library version

Co-authored-by: Halvard Mørstad <33620089+halvardssm@users.noreply.github.com>

asantos00 committed:
Typo in test name

Co-authored-by: Halvard Mørstad <33620089+halvardssm@users.noreply.github.com>

asantos00 committed:
fix: use errorMessages object, differentiate jwt errors

asantos00 committed:
Merge branch 'update-djwt-2-0' of github.com:asantos00/oak-middleware-jwt into update-djwt-2-0

asantos00 commented on src/jwtMiddleware.ts:
```diff
@@ -88,53 +94,60 @@ export const jwtMiddleware = <
-        }
+    const onUnauthorized = async (
+      jwtValidation: Error,
+      isJwtValidationError: boolean = false,
```
I've called this `isJwtValidationError` because I didn't middleware users want it to be too much coupled with the underlying library (djwt). But I'm not sure I fully understood your intentions and thus I haven't changed much of the way errors were handled here

halvardssm replied:
so what I meant with this is that if this arg is true, the message on line 121, should be the actual error message rather than one of the custom messages, i.e. `isJwtValidationError ? jwtValidation.message :  customMessages?.ERROR_INVALID_AUTH,`

asantos00 commented on README.md:
```diff
@@ -57,24 +57,55 @@ Oak middleware for JWT
 
-* key: string; // See the djwt module for Validation options
-* algorithm: Algorithm | Algorithm[]; // See the djwt module for Validation options
+* secret: string; // See the djwt module for Validation options
```
Happy to revert this, of course. It was just for you to get what I was trying to say.

halvardssm replied:
I looked at the djwt module, and it seems like they use the terms interchangeably, but that the methods use `key` as the variable name. Am I missing something in regards to why you want to change this? I am fine with changing this, I am just curious to why 😄

asantos00 commented on README.md:
```diff
@@ -57,24 +57,55 @@ Oak middleware for JWT
+* onSuccess?: OnSuccessHandler; // Optional callback for successfull validation, passes the Context and the Payload object from djwt module
+* onFailure?: OnFailureHandler; // Optional callback for unsuccessfull validation, passes the Context and the Error encountered while validating the jwt
+
+## Migrating from v1.0.0
```
Do you think this makes sense?

halvardssm replied:
Looks good to me, the only thing I am wondering about is the `key`->`secret` change

asantos00 commented on src/jwtMiddleware.ts:
```diff
@@ -87,54 +94,68 @@ export const jwtMiddleware = <
+  onSuccess = () => {},
+  onFailure = () => true,
 }: JwtMiddlewareOptions): T => {
+  if (typeof key === undefined) {
+    console.warn(
+      `oak-middleware-jwt: Property 'key' is going to be deprecated, you should now use 'secret' instead.`,
+    );
+  }
```
Also added this since this is breaking, might not make sense because the type itself explains this 🤷‍♂️

halvardssm replied:
we can skip this as long as we have the deprecated notice for the type

asantos00 commented:
Got my changes done. I'd like you to clarify a few things:
- If you want the "error from djwt" flag send into the `onFailure` callback
- If you want to keep/remove the deprecation warning log
- If you want to keep the "migrating from version 1.0.0" section on README

halvardssm commented on .github/workflows/ci.yml:
```diff
@@ -1,7 +1,7 @@
 
 env:
-  DENO_VERSION: 1.4.6
+  DENO_VERSION: 1.6.0
```
Let's go directly to 1.6.3 since that is the latest at the time of writing.
```suggestion
  DENO_VERSION: 1.6.3
```

halvardssm reviewed (commented):
Just some small comments, but I think we are almost good to go 👍🏻

halvardssm commented on src/jwtMiddleware.ts:
```diff
@@ -38,19 +42,22 @@ export interface JwtMiddlewareOptions extends Partial<Validation> {
   /** See the djwt module for Validation options */
-  key: string;
-  algorithm: Algorithm | Algorithm[];
+  key?: string; // Deprecated
```
```suggestion
  /** @deprecated */
  key?: string; 
```

asantos00 committed:
Update deno version on CI

Co-authored-by: Halvard Mørstad <33620089+halvardssm@users.noreply.github.com>

asantos00 commented:
Lol 🤦 I was looking at the "secret" they're using as an example of a key, and not to the name of the argument itself. You are completely right, I'm reverting this 😄 
![image](https://user-images.githubusercontent.com/2593480/103383199-c3569980-4ae9-11eb-9a52-93924a0163e9.png)

asantos00 committed:
fix: get key back instead of secret

asantos00 committed:
Merge branch 'update-djwt-2-0' of github.com:asantos00/oak-middleware-jwt into update-djwt-2-0

asantos00 committed:
test: run server on random port

The main reason for this is because if a test fails (or tests run in
parallel) the server fails to start because the port is in use.

halvardssm commented on README.md:
```diff
@@ -57,24 +57,53 @@ Oak middleware for JWT
+
+- Change the onFailure and onSuccess callbacks.
+  - `onSuccess` gets an object of type `Payload` as a second argument (check https://github.com/timonson/djwt#decode)
+  - `onFailure` gets an object of type `Error` as a second argument
```
```suggestion
  - `onFailure` gets an object of type `Error` as a second argument, should return `true` if the error should be thrown instead of returning as a response.
```

halvardssm reviewed (approved):
Looks good to me! Thank you for your hard work! Just let me know if you are completely finished and I will merge it 👍🏻

asantos00 commented:
Thank you for the collaboration, feedback and openness to changes 🙏
Feel free to merge it whenever it makes sense. It would be really cool to have a tag published too 👌

halvardssm committed:
Update README.md

asantos00 commented:
Do you think you can publish the new version on deno.land/x or nest.land?
Thanks!

halvardssm commented:
@asantos00 Should be uploaded to nest now! I dont think deno.land/x is a place to pubish an oak middleware (to not clutter the x space). But if you want it available on other "package managers" please let me know which one! :)
